nemo/collections/nlp/nm/losses/sgd_loss.py

```diff
@@ -0,0 +1,219 @@
+        # Shape: (batch_size, max_num_noncat_slots, 3).
+        if sum(non_cat_slot_status_mask) == 0:
+            logging.warning(f'No active non-categorical slots in the batch.')
+            noncat_slot_status_loss = self._cross_entropy(
+                logit_noncat_slot_status.view(-1, 3), torch.argmax(logit_noncat_slot_status.view(-1, 3), dim=-1)
+            )
+        else:
+            noncat_slot_status_loss = self._cross_entropy(
+                logit_noncat_slot_status.view(-1, 3)[non_cat_slot_status_mask],
+                noncategorical_slot_status.view(-1)[non_cat_slot_status_mask],
+            )
+
+        # Non-categorical slot spans.
+        # Shape: (batch_size, max_num_noncat_slots, max_num_tokens).n
+        max_num_tokens = logit_noncat_slot_start.size()[-1]
+        # Zero out losses for non-categorical slot spans when the slot status is not active.
+        # changed here
+        non_cat_slot_value_mask = (noncategorical_slot_status == STATUS_ACTIVE).view(-1)
+        # non_cat_slot_value_mask = (noncategorical_slot_status > -1 ).view(-1)
+        # to handle cases with no active categorical slot value
+        if sum(non_cat_slot_value_mask) == 0:
+            logging.warning(f'No active values for non-categorical slots in the batch.')
+            span_start_loss = self._cross_entropy(
+                logit_noncat_slot_start.view(-1, max_num_tokens),
+                torch.argmax(logit_noncat_slot_start.view(-1, max_num_tokens), dim=-1),
+            )
+            span_end_loss = self._cross_entropy(
+                logit_noncat_slot_end.view(-1, max_num_tokens),
+                torch.argmax(logit_noncat_slot_end.view(-1, max_num_tokens), dim=-1),
+            )
+        else:
+            noncat_slot_start_active_logits = logit_noncat_slot_start.view(-1, max_num_tokens)[non_cat_slot_value_mask]
+            noncat_slot_start_active_labels = noncategorical_slot_value_start.view(-1)[non_cat_slot_value_mask]
+            span_start_loss = self._cross_entropy(noncat_slot_start_active_logits, noncat_slot_start_active_labels)
+
+            noncat_slot_end_active_logits = logit_noncat_slot_end.view(-1, max_num_tokens)[non_cat_slot_value_mask]
+            noncat_slot_end_active_labels = noncategorical_slot_value_end.view(-1)[non_cat_slot_value_mask]
+            span_end_loss = self._cross_entropy(noncat_slot_end_active_logits, noncat_slot_end_active_labels)
+
+        losses = {
+            "intent_loss": intent_loss,
+            "requested_slot_loss": requested_slot_loss,
+            "cat_slot_status_loss": cat_slot_status_loss,
+            "cat_slot_value_loss": cat_slot_value_loss,
+            "noncat_slot_status_loss": noncat_slot_status_loss,
+            "span_start_loss": span_start_loss,
+            "span_end_loss": span_end_loss,
+        }
+
+        total_loss = sum(losses.values())
+        if self.reduction == 'mean':
+            total_loss = total_loss / len(losses)
```

Comment:
shouldn't we still (additionally) divide it by the batch_size??

Reply:
no, mean reduction will take care of it